File: go/turbodata/multi_reader_test.go
package turbodata

import (
	"bytes"
	"errors"
	"fmt"
	"io"
	"testing"
)

// buildOneTopic builds a single-topic file with one message per timestamp.
// Message payload is "<topic>@<ts>" so callers can assert provenance.
func buildOneTopic(t *testing.T, topic string, tss ...int64) []byte {
	t.Helper()
	return buildFile(t, func(w *Writer) {
		mustOpenTopics(t, w, []string{topic}, []map[string]any{{}})
		for _, ts := range tss {
			mustWriteMessage(t, w, topic, []byte(fmt.Sprintf("%s@%d", topic, ts)), ts)
		}
		mustCloseTopic(t, w)
		mustClose(t, w)
	})
}

// collectMulti drains a MultiReader into a slice of testMsgs.
func collectMulti(t *testing.T, mr *MultiReader, opts ...ReadOption) []testMsg {
	t.Helper()
	it, err := mr.ReadMessages(opts...)
	if err != nil {
		t.Fatalf("ReadMessages: %v", err)
	}
	rb := NewReusableBuffer()
	var msgs []testMsg
	for {
		ts, name, err := it.NextInto(rb)
		if errors.Is(err, io.EOF) {
			break
		}
		if err != nil {
			t.Fatalf("NextInto: %v", err)
		}
		d := make([]byte, len(rb.Data))
		copy(d, rb.Data)
		msgs = append(msgs, testMsg{ts, name, d})
	}
	return msgs
}

// TestMultiReadTimeSplit: two files holding the same topic over disjoint time
// ranges merge into one fully ordered stream, forward and reverse.
func TestMultiReadTimeSplit(t *testing.T) {
	a := buildOneTopic(t, "cam", 10, 30, 50)
	b := buildOneTopic(t, "cam", 20, 40, 60)
	mr, err := NewMultiReader(
		NewReader(bytes.NewReader(a)),
		NewReader(bytes.NewReader(b)),
	)
	if err != nil {
		t.Fatalf("NewMultiReader: %v", err)
	}

	msgs := collectMulti(t, mr)
	wantTs := []int64{10, 20, 30, 40, 50, 60}
	if len(msgs) != len(wantTs) {
		t.Fatalf("forward count got %d want %d (%+v)", len(msgs), len(wantTs), msgs)
	}
	for i, ts := range wantTs {
		if msgs[i].ts != ts || msgs[i].name != "cam" {
			t.Errorf("forward[%d] got (%d,%q) want (%d,cam)", i, msgs[i].ts, msgs[i].name, ts)
		}
	}

	rev := collectMulti(t, mr, WithOrder(ReverseTimeOrder))
	wantRev := []int64{60, 50, 40, 30, 20, 10}
	if len(rev) != len(wantRev) {
		t.Fatalf("reverse count got %d want %d (%+v)", len(rev), len(wantRev), rev)
	}
	for i, ts := range wantRev {
		if rev[i].ts != ts {
			t.Errorf("reverse[%d] ts got %d want %d", i, rev[i].ts, ts)
		}
	}
}

// TestMultiReadAugmentation: two files with disjoint topics over overlapping
// time merge into the time-ordered union of both topics.
func TestMultiReadAugmentation(t *testing.T) {
	a := buildOneTopic(t, "lidar", 10, 20, 30)
	b := buildOneTopic(t, "radar", 15, 25, 35)
	mr, err := NewMultiReader(
		NewReader(bytes.NewReader(a)),
		NewReader(bytes.NewReader(b)),
	)
	if err != nil {
		t.Fatalf("NewMultiReader: %v", err)
	}

	msgs := collectMulti(t, mr)
	type tn struct {
		ts   int64
		name string
	}
	want := []tn{{10, "lidar"}, {15, "radar"}, {20, "lidar"}, {25, "radar"}, {30, "lidar"}, {35, "radar"}}
	if len(msgs) != len(want) {
		t.Fatalf("count got %d want %d (%+v)", len(msgs), len(want), msgs)
	}
	for i, w := range want {
		if msgs[i].ts != w.ts || msgs[i].name != w.name {
			t.Errorf("[%d] got (%d,%q) want (%d,%q)", i, msgs[i].ts, msgs[i].name, w.ts, w.name)
		}
	}
}

// TestMultiReadUnionVsSplit: a shared in-file name unions across files, while
// a per-Reader remap keeps the colliding name distinct.
func TestMultiReadUnionVsSplit(t *testing.T) {
	a := buildOneTopic(t, "cam", 10, 30)
	b := buildOneTopic(t, "cam", 20, 40)

	union, err := NewMultiReader(
		NewReader(bytes.NewReader(a)),
		NewReader(bytes.NewReader(b)),
	)
	if err != nil {
		t.Fatalf("NewMultiReader union: %v", err)
	}
	um := collectMulti(t, union)
	if len(um) != 4 {
		t.Fatalf("union count got %d want 4", len(um))
	}
	for i, m := range um {
		if m.name != "cam" {
			t.Errorf("union[%d] name got %q want cam", i, m.name)
		}
	}

	split, err := NewMultiReader(
		NewReader(bytes.NewReader(a)),
		NewReader(bytes.NewReader(b), WithTopicRemap(map[string]string{"cam": "cam_b"})),
	)
	if err != nil {
		t.Fatalf("NewMultiReader split: %v", err)
	}
	sm := collectMulti(t, split)
	names := map[string]int{}
	for _, m := range sm {
		names[m.name]++
	}
	if names["cam"] != 2 || names["cam_b"] != 2 {
		t.Errorf("split names got %v want cam:2 cam_b:2", names)
	}
}

// buildVideoFile builds a single video-topic file from the given frames.
func buildVideoFile(t *testing.T, topic string, frames []frame) []byte {
	t.Helper()
	return buildFile(t, func(w *Writer) {
		mustOpenTopics(t, w, []string{topic}, []map[string]any{{}}, WithVideoTopic())
		writeVideo(t, w, topic, frames)
		mustCloseTopic(t, w)
		mustClose(t, w)
	})
}

// TestMultiSampleLatestFloor: the same non-video topic split across two files
// resolves each timestamp to the latest floor across the union.
func TestMultiSampleLatestFloor(t *testing.T) {
	a := buildOneTopic(t, "cam", 10, 20)
	b := buildOneTopic(t, "cam", 30, 40)
	mr, err := NewMultiReader(
		NewReader(bytes.NewReader(a)),
		NewReader(bytes.NewReader(b)),
	)
	if err != nil {
		t.Fatalf("NewMultiReader: %v", err)
	}

	out, err := mr.Sample([]SampleQuery{
		{Topic: "cam", Timestamps: []int64{5, 15, 35, 50}},
	})
	if err != nil {
		t.Fatalf("Sample: %v", err)
	}
	want := []SampleResult{
		{Found: false},
		{Found: true, Timestamp: 10, Data: []byte("cam@10")},
		{Found: true, Timestamp: 30, Data: []byte("cam@30")},
		{Found: true, Timestamp: 40, Data: []byte("cam@40")},
	}
	assertResultsEqual(t, out[0], want)
}

// TestMultiSampleVideoTimeDisjoint: a video topic spread across two
// time-disjoint files samples correctly, with each picked cell carrying its
// reader's GOP prefix (ResetDecoder=true at the switch boundary).
func TestMultiSampleVideoTimeDisjoint(t *testing.T) {
	aFrames := []frame{
		{ts: 10, isKeyFrame: true, data: []byte("K10")},
		{ts: 20, isKeyFrame: false, data: []byte("P20")},
		{ts: 30, isKeyFrame: false, data: []byte("P30")},
	}
	bFrames := []frame{
		{ts: 100, isKeyFrame: true, data: []byte("K100")},
		{ts: 110, isKeyFrame: false, data: []byte("P110")},
		{ts: 120, isKeyFrame: false, data: []byte("P120")},
	}
	a := buildVideoFile(t, "cam", aFrames)
	b := buildVideoFile(t, "cam", bFrames)
	mr, err := NewMultiReader(
		NewReader(bytes.NewReader(a)),
		NewReader(bytes.NewReader(b)),
	)
	if err != nil {
		t.Fatalf("NewMultiReader: %v", err)
	}

	out, err := mr.Sample([]SampleQuery{
		{Topic: "cam", Timestamps: []int64{25, 115}},
	}, WithSampleVideoDecodable())
	if err != nil {
		t.Fatalf("Sample: %v", err)
	}
	row := out[0]
	if len(row) != 2 {
		t.Fatalf("expected 2 results, got %d", len(row))
	}

	// T=25 -> file A, floor P20, full GOP prefix [K10, P20].
	if !row[0].Found || !row[0].IsVideo || !row[0].ResetDecoder || row[0].Timestamp != 20 {
		t.Errorf("row[0]: got Found=%v IsVideo=%v Reset=%v ts=%d; want true,true,true,20",
			row[0].Found, row[0].IsVideo, row[0].ResetDecoder, row[0].Timestamp)
	}
	assertFramesMatch(t, row[0].Frames, []frame{aFrames[0], aFrames[1]})

	// T=115 -> file B wins (latest floor), first Found cell of B's row so
	// ResetDecoder=true with B's GOP prefix [K100, P110].
	if !row[1].Found || !row[1].IsVideo || !row[1].ResetDecoder || row[1].Timestamp != 110 {
		t.Errorf("row[1]: got Found=%v IsVideo=%v Reset=%v ts=%d; want true,true,true,110",
			row[1].Found, row[1].IsVideo, row[1].ResetDecoder, row[1].Timestamp)
	}
	assertFramesMatch(t, row[1].Frames, []frame{bFrames[0], bFrames[1]})
}

// TestMultiSampleVideoOverlapError: overlapping video sources are rejected
// under WithSampleVideoDecodable, but allowed (plain floor) without it.
func TestMultiSampleVideoOverlapError(t *testing.T) {
	aFrames := []frame{
		{ts: 10, isKeyFrame: true, data: []byte("K10")},
		{ts: 30, isKeyFrame: false, data: []byte("P30")},
		{ts: 50, isKeyFrame: false, data: []byte("P50")},
	}
	bFrames := []frame{
		{ts: 40, isKeyFrame: true, data: []byte("K40")},
		{ts: 60, isKeyFrame: false, data: []byte("P60")},
		{ts: 80, isKeyFrame: false, data: []byte("P80")},
	}
	a := buildVideoFile(t, "cam", aFrames)
	b := buildVideoFile(t, "cam", bFrames)
	mr, err := NewMultiReader(
		NewReader(bytes.NewReader(a)),
		NewReader(bytes.NewReader(b)),
	)
	if err != nil {
		t.Fatalf("NewMultiReader: %v", err)
	}

	// Overlapping [10,50] and [40,80] under the decodable option is rejected.
	_, err = mr.Sample([]SampleQuery{
		{Topic: "cam", Timestamps: []int64{45}},
	}, WithSampleVideoDecodable())
	if !errors.Is(err, ErrVideoSourcesOverlap) {
		t.Errorf("expected ErrVideoSourcesOverlap, got %v", err)
	}

	// Without the decodable option, overlap is fine: plain floor semantics.
	out, err := mr.Sample([]SampleQuery{
		{Topic: "cam", Timestamps: []int64{45}},
	})
	if err != nil {
		t.Fatalf("Sample (plain): %v", err)
	}
	// Floor of 45 across both: A=P30(30), B=K40(40) -> latest 40.
	got := out[0][0]
	if !got.Found || got.Timestamp != 40 || string(got.Data) != "K40" {
		t.Errorf("plain floor: got Found=%v ts=%d data=%q; want true,40,K40", got.Found, got.Timestamp, got.Data)
	}
}

// buildVideoPlusTopic builds a file with one video topic and one plain topic.
func buildVideoPlusTopic(t *testing.T, videoTopic string, vframes []frame, plainTopic string, plainTs []int64) []byte {
	t.Helper()
	return buildFile(t, func(w *Writer) {
		mustOpenTopics(t, w, []string{videoTopic}, []map[string]any{{}}, WithVideoTopic())
		writeVideo(t, w, videoTopic, vframes)
		mustCloseTopic(t, w)
		mustOpenTopics(t, w, []string{plainTopic}, []map[string]any{{}})
		for _, ts := range plainTs {
			mustWriteMessage(t, w, plainTopic, []byte(fmt.Sprintf("%s@%d", plainTopic, ts)), ts)
		}
		mustCloseTopic(t, w)
		mustClose(t, w)
	})
}

// TestMultiReadVideoTimeDisjoint: a video topic split across two time-disjoint
// files merges fine under WithVideoDecodable.
func TestMultiReadVideoTimeDisjoint(t *testing.T) {
	aFrames := []frame{
		{ts: 10, isKeyFrame: true, data: []byte("K10")},
		{ts: 20, isKeyFrame: false, data: []byte("P20")},
		{ts: 30, isKeyFrame: false, data: []byte("P30")},
	}
	bFrames := []frame{
		{ts: 100, isKeyFrame: true, data: []byte("K100")},
		{ts: 110, isKeyFrame: false, data: []byte("P110")},
		{ts: 120, isKeyFrame: false, data: []byte("P120")},
	}
	mr, err := NewMultiReader(
		NewReader(bytes.NewReader(buildVideoFile(t, "cam", aFrames))),
		NewReader(bytes.NewReader(buildVideoFile(t, "cam", bFrames))),
	)
	if err != nil {
		t.Fatalf("NewMultiReader: %v", err)
	}

	msgs := collectMulti(t, mr, WithVideoDecodable())
	wantTs := []int64{10, 20, 30, 100, 110, 120}
	if len(msgs) != len(wantTs) {
		t.Fatalf("count got %d want %d (%+v)", len(msgs), len(wantTs), msgs)
	}
	for i, ts := range wantTs {
		if msgs[i].ts != ts || msgs[i].name != "cam" {
			t.Errorf("[%d] got (%d,%q) want (%d,cam)", i, msgs[i].ts, msgs[i].name, ts)
		}
	}
}

// TestMultiReadVideoOverlapError: under WithVideoDecodable, overlapping video
// sources are rejected (merging them would interleave GOP chains); without the
// option the merge is allowed.
func TestMultiReadVideoOverlapError(t *testing.T) {
	aFrames := []frame{
		{ts: 10, isKeyFrame: true, data: []byte("K10")},
		{ts: 30, isKeyFrame: false, data: []byte("P30")},
		{ts: 50, isKeyFrame: false, data: []byte("P50")},
	}
	bFrames := []frame{
		{ts: 40, isKeyFrame: true, data: []byte("K40")},
		{ts: 60, isKeyFrame: false, data: []byte("P60")},
		{ts: 80, isKeyFrame: false, data: []byte("P80")},
	}
	mr, err := NewMultiReader(
		NewReader(bytes.NewReader(buildVideoFile(t, "cam", aFrames))),
		NewReader(bytes.NewReader(buildVideoFile(t, "cam", bFrames))),
	)
	if err != nil {
		t.Fatalf("NewMultiReader: %v", err)
	}

	// Overlapping [10,50] and [40,80] under WithVideoDecodable is rejected.
	if _, err := mr.ReadMessages(WithVideoDecodable()); !errors.Is(err, ErrVideoSourcesOverlap) {
		t.Errorf("expected ErrVideoSourcesOverlap, got %v", err)
	}

	// Without the option, the merge is allowed (no decode contract).
	if _, err := mr.ReadMessages(); err != nil {
		t.Errorf("plain ReadMessages should not error, got %v", err)
	}
}

// TestMultiReadVideoOverlapOutOfScope: an overlapping video topic that is
// filtered out via WithTopicNames must not trigger the disjoint check.
func TestMultiReadVideoOverlapOutOfScope(t *testing.T) {
	aFrames := []frame{
		{ts: 10, isKeyFrame: true, data: []byte("K10")},
		{ts: 50, isKeyFrame: false, data: []byte("P50")},
	}
	bFrames := []frame{
		{ts: 40, isKeyFrame: true, data: []byte("K40")},
		{ts: 80, isKeyFrame: false, data: []byte("P80")},
	}
	mr, err := NewMultiReader(
		NewReader(bytes.NewReader(buildVideoPlusTopic(t, "cam", aFrames, "imu", []int64{15, 25}))),
		NewReader(bytes.NewReader(buildVideoPlusTopic(t, "cam", bFrames, "imu", []int64{45, 55}))),
	)
	if err != nil {
		t.Fatalf("NewMultiReader: %v", err)
	}

	// "cam" overlaps, but it is out of scope: only "imu" is requested.
	msgs := collectMulti(t, mr, WithVideoDecodable(), WithTopicNames([]string{"imu"}))
	wantTs := []int64{15, 25, 45, 55}
	if len(msgs) != len(wantTs) {
		t.Fatalf("count got %d want %d (%+v)", len(msgs), len(wantTs), msgs)
	}
	for i, ts := range wantTs {
		if msgs[i].ts != ts || msgs[i].name != "imu" {
			t.Errorf("[%d] got (%d,%q) want (%d,imu)", i, msgs[i].ts, msgs[i].name, ts)
		}
	}

	// Bringing "cam" into scope re-triggers the overlap error.
	if _, err := mr.ReadMessages(WithVideoDecodable(), WithTopicNames([]string{"cam"})); !errors.Is(err, ErrVideoSourcesOverlap) {
		t.Errorf("in-scope cam: expected ErrVideoSourcesOverlap, got %v", err)
	}
}

File: ts/test/multi_reader.test.ts
// MultiReader tests. Mirrors go/multi_reader_test.go and
// py/tests/test_multi_reader.py.
//
// The TS SDK has no Writer, so these are driven off the Go-produced fixtures.
// Two readers over time-disjoint /sensor files (single-topic-uncompressed.td
// at ts 1.00M..1.029M and single-topic-compressed.td at ts 2.00M..2.029M)
// stand in for the synthetic "same topic, disjoint time" files the Go/Python
// tests build in memory. The /cam video fixtures (all ts 10..60) exercise the
// overlap path; remap splits them apart for the disjoint case.

import { createHash } from "node:crypto";

import { describe, expect, it } from "vitest";

import {
  MultiReader,
  Reader,
  SampleValidationError,
  VideoSourcesOverlapError,
  type Message,
  type ReadOptions,
  type SampleResult,
} from "../src/index.js";

import { BytesReadSource, readFixture } from "./helpers.js";

const SENSOR_A = "single-topic-uncompressed.td"; // /sensor, ts 1000000..1029000
const SENSOR_B = "single-topic-compressed.td"; // /sensor, ts 2000000..2029000
const MULTI = "multi-topic-compressed.td"; // /cam, /imu, /lidar, ts ~3000000
const VIDEO_A = "video-single-gop.td"; // /cam, ts 10..60
const VIDEO_B = "video-multi-gop-chunks.td"; // /cam, ts 10..60

function hex(data: Uint8Array): string {
  return createHash("sha256").update(data).digest("hex");
}

function reader(name: string, opts = {}): Reader {
  return new Reader(new BytesReadSource(readFixture(name)), opts);
}

interface Row {
  ts: bigint;
  name: string;
  data: string;
}

async function drain(
  it: AsyncIterableIterator<Message>,
): Promise<Row[]> {
  const out: Row[] = [];
  for await (const msg of it) {
    out.push({ ts: msg.timestamp, name: msg.topicName, data: hex(msg.data) });
  }
  return out;
}

function collectMulti(mr: MultiReader, opts: ReadOptions = {}): Promise<Row[]> {
  return drain(mr.readMessages({ copy: true, ...opts }));
}

describe("MultiReader construction", () => {
  it("requires at least one reader", () => {
    expect(() => new MultiReader([])).toThrow(/at least one reader/);
  });
});

describe("MultiReader readMessages", () => {
  it("merges the same topic over disjoint time into one ordered stream", async () => {
    const mr = new MultiReader([reader(SENSOR_A), reader(SENSOR_B)]);
    const rows = await collectMulti(mr);

    expect(rows.length).toBe(60);
    expect(rows.every((r) => r.name === "/sensor")).toBe(true);
    // Strictly ascending across the merge boundary.
    for (let i = 1; i < rows.length; i++) {
      expect(rows[i]!.ts > rows[i - 1]!.ts).toBe(true);
    }
    // First half from A (~1M), second half from B (~2M).
    expect(rows[0]!.ts).toBe(1000000n);
    expect(rows[29]!.ts).toBe(1029000n);
    expect(rows[30]!.ts).toBe(2000000n);
    expect(rows[59]!.ts).toBe(2029000n);
  });

  it("merges disjoint topics into the time-ordered union", async () => {
    const mr = new MultiReader([reader(SENSOR_A), reader(MULTI)]);
    const merged = await collectMulti(mr);

    // Ground truth: read each reader alone and merge by timestamp in JS.
    const a = await drain(reader(SENSOR_A).readMessages({ copy: true }));
    const b = await drain(reader(MULTI).readMessages({ copy: true }));
    const expected = [...a, ...b].sort((x, y) =>
      x.ts < y.ts ? -1 : x.ts > y.ts ? 1 : 0,
    );

    expect(merged.length).toBe(a.length + b.length);
    expect(merged).toEqual(expected);
  });

  it("unions a shared name but a per-reader remap keeps it distinct", async () => {
    const union = new MultiReader([reader(SENSOR_A), reader(SENSOR_B)]);
    const um = await collectMulti(union);
    expect(um.length).toBe(60);
    expect(um.every((r) => r.name === "/sensor")).toBe(true);

    const split = new MultiReader([
      reader(SENSOR_A),
      reader(SENSOR_B, { topicRemap: { "/sensor": "/sensor_b" } }),
    ]);
    const sm = await collectMulti(split);
    const names: Record<string, number> = {};
    for (const r of sm) {
      names[r.name] = (names[r.name] ?? 0) + 1;
    }
    expect(names).toEqual({ "/sensor": 30, "/sensor_b": 30 });
  });

  it("merges in reverse-time order", async () => {
    const mr = new MultiReader([reader(SENSOR_A), reader(SENSOR_B)]);
    const rows = await collectMulti(mr, { order: "reverse-time" });

    expect(rows.length).toBe(60);
    expect(rows[0]!.ts).toBe(2029000n);
    expect(rows[59]!.ts).toBe(1000000n);
    for (let i = 1; i < rows.length; i++) {
      expect(rows[i]!.ts < rows[i - 1]!.ts).toBe(true);
    }
  });

  it("forwards topicNames filtering (by exposed name) to each reader", async () => {
    const mr = new MultiReader([reader(SENSOR_A), reader(MULTI)]);
    const rows = await collectMulti(mr, { topicNames: ["/imu"] });
    expect(rows.length).toBeGreaterThan(0);
    expect(rows.every((r) => r.name === "/imu")).toBe(true);
  });
});

describe("MultiReader sample", () => {
  it("resolves each timestamp to the latest floor across the union", async () => {
    const mr = new MultiReader([reader(SENSOR_A), reader(SENSOR_B)]);
    const ts = [1000000n, 1500000n, 2000000n, 2050000n];
    const out = await mr.sample([{ topic: "/sensor", timestamps: ts }]);
    const row = out[0]!;

    expect(row.map((r) => r.found)).toEqual([true, true, true, true]);
    // 1.00M -> A exact; 1.5M -> A's last (no B floor); 2.00M -> B exact wins
    // over A's last; 2.05M -> B's last wins.
    expect(row.map((r) => r.timestamp)).toEqual([
      1000000n,
      1029000n,
      2000000n,
      2029000n,
    ]);

    // Provenance: the B-winning cells carry B's data, not A's.
    const bAlone = await reader(SENSOR_B).sample([
      { topic: "/sensor", timestamps: [2000000n, 2050000n] },
    ]);
    expect(hex(row[2]!.data)).toBe(hex(bAlone[0]![0]!.data));
    expect(hex(row[3]!.data)).toBe(hex(bAlone[0]![1]!.data));
  });

  it("returns not-found before any reader's first message", async () => {
    const mr = new MultiReader([reader(SENSOR_A), reader(SENSOR_B)]);
    const out = await mr.sample([{ topic: "/sensor", timestamps: [1n] }]);
    expect(out[0]![0]!.found).toBe(false);
  });

  it("rejects a topic held by no reader", async () => {
    const mr = new MultiReader([reader(SENSOR_A)]);
    await expect(
      mr.sample([{ topic: "/nope", timestamps: [1n] }]),
    ).rejects.toThrow(SampleValidationError);
  });

  it("rejects non-strictly-increasing timestamps", async () => {
    const mr = new MultiReader([reader(SENSOR_A)]);
    await expect(
      mr.sample([{ topic: "/sensor", timestamps: [10n, 10n] }]),
    ).rejects.toThrow(SampleValidationError);
  });
});

describe("MultiReader video", () => {
  it("rejects overlapping video sources under videoDecodable (read)", async () => {
    const mr = new MultiReader([reader(VIDEO_A), reader(VIDEO_B)]);
    // /cam is video in both files over the same ts range [10,60] -> overlap.
    await expect(collectMulti(mr, { videoDecodable: true })).rejects.toThrow(
      VideoSourcesOverlapError,
    );

    // Without the option the merge is allowed (plain interleave, no contract).
    const plain = await collectMulti(new MultiReader([reader(VIDEO_A), reader(VIDEO_B)]));
    expect(plain.length).toBe(12);
    expect(plain.every((r) => r.name === "/cam")).toBe(true);
  });

  it("rejects overlapping video sources under videoDecodable (sample)", async () => {
    const mr = new MultiReader([reader(VIDEO_A), reader(VIDEO_B)]);
    await expect(
      mr.sample([{ topic: "/cam", timestamps: [35n] }], { videoDecodable: true }),
    ).rejects.toThrow(VideoSourcesOverlapError);

    // Without the decodable option, plain floor semantics across the union.
    const out = await new MultiReader([reader(VIDEO_A), reader(VIDEO_B)]).sample([
      { topic: "/cam", timestamps: [35n] },
    ]);
    expect(out[0]![0]!.found).toBe(true);
  });

  it("does not check an overlapping video topic that is out of scope", async () => {
    const mr = new MultiReader([reader(VIDEO_A), reader(VIDEO_B)]);
    // /cam overlaps, but only a non-existent topic is requested: no error,
    // empty stream.
    const rows = await collectMulti(mr, {
      videoDecodable: true,
      topicNames: ["/not-cam"],
    });
    expect(rows.length).toBe(0);

    // Bringing /cam into scope re-triggers the overlap error.
    await expect(
      collectMulti(new MultiReader([reader(VIDEO_A), reader(VIDEO_B)]), {
        videoDecodable: true,
        topicNames: ["/cam"],
      }),
    ).rejects.toThrow(VideoSourcesOverlapError);
  });

  it("allows decodable video when remap splits the topics apart", async () => {
    // Distinct exposed names => each video topic is held by one reader, so the
    // disjoint check is a no-op and decodable read/sample succeed.
    const mr = new MultiReader([
      reader(VIDEO_A, { topicRemap: { "/cam": "/camA" } }),
      reader(VIDEO_B, { topicRemap: { "/cam": "/camB" } }),
    ]);

    const rows = await collectMulti(mr, { videoDecodable: true });
    expect(rows.length).toBe(12);
    const names = new Set(rows.map((r) => r.name));
    expect(names).toEqual(new Set(["/camA", "/camB"]));

    const out = await mr.sample(
      [
        { topic: "/camA", timestamps: [35n] },
        { topic: "/camB", timestamps: [35n] },
      ],
      { videoDecodable: true },
    );
    for (const row of out as SampleResult[][]) {
      expect(row[0]!.found).toBe(true);
      expect(row[0]!.isVideo).toBe(true);
      expect(row[0]!.frames.length).toBeGreaterThan(0);
    }
  });
});
